syntax = "proto2";

message Underlying {

    message Asset {
        optional string inst_id = 1;
        optional double weight = 2;
    }

    enum UnderlyingType {
        Single = 0;
        FixedWeightBasket = 1;
        WorstOfBasket = 2;
        BestOfBasket = 3;
    }

    optional UnderlyingType type = 1;
    repeated Asset assets = 2;

}

message DrivativeTraits {

    enum OptionType {
        Call = 0;
        Put = 1;
    }

    enum OptionStyle {
        European = 0;
        American = 1;
    }

    optional OptionType option_type = 1;
    optional OptionStyle option_style = 2;
    optional double strike = 3;
    optional int64 exp_date = 4;
    optional double multiplier = 5;
}


message Instrument {

    enum InstType {
        STK = 0;
        FUT = 1;
        OPT = 2;
        FOT = 3;
        IDX = 4;
        CASH = 5;
        ETF = 6;
        CBO = 7;
    }

    //int64 inst_id = 1; // unique, generated by system
    optional string inst_id = 2; // uniqie symbol + '@' + primary_exch_id
    optional string symbol = 3;
    optional string name = 4;
    optional InstType type = 5;
    optional string primary_exch_id = 6;
    repeated string exch_ids = 7;
    optional string ccy_id = 8;
    optional string sector = 9;
    optional string industry = 10;
    optional double margin = 11;
    optional double tick_size = 12;

    //derivatives
    optional Underlying underlying = 21;
    optional DrivativeTraits derivative = 22;

    //alt sym / ids
    map<string, string> alt_symbols = 31; //map<provider_id, alt_symbol>, e.g. IB -> 5
    map<string, string> alt_ids = 32; //e.g. RIC -> 0005.HK
    map<string, string> alt_sectors = 33;
    map<string, string> alt_industries = 34;
}

message Exchange {
    optional string exch_id = 1; //uniqie
    optional string name = 2;
    optional string country_id = 3;
    optional string trading_hours_id = 4;
    optional string holidays_id = 5;

    map<string, string> alt_ids = 6;
}

message Country {
    optional string country_id = 1; //uniqie
    optional string name = 2;

    optional string holidays_id = 3;
}

message Currency {
    optional string ccy_id = 1; //uniqie
    optional string name = 2;
}

message HolidaySeries {
    message Holiday {
        optional int64 trading_date = 1;
        optional int64 start_date = 2;
        optional int64 start_time = 3;
        optional int64 end_date = 4;
        optional int64 end_time = 5;
        optional Type type = 6;
        optional string desc = 7;

        enum Type {
            FullDay = 0;
            LateOpen = 1;
            EarlyClose = 2;
            Replace = 3;
            Modify = 4;
        }
    }

    optional string holidays_id = 1;
    repeated Holiday holidays = 2;
}

message TradingHours {
    message Session {
        optional WeekDay start_weekdate = 1;
        optional int64 start_time = 2;
        optional WeekDay end_weekdate = 3;
        optional int64 end_time = 4;
        optional bool eod = 5;

        enum WeekDay {
            Sunday = 0;
            Monday = 1;
            Tuesday = 2;
            Wednesday = 3;
            Thursday = 4;
            Friday = 5;
            Saturday = 6;
        }
    }

    optional string trading_hours_id = 1;
    optional string timezone_id = 2;
    repeated Session sessions = 3;
}

message TimeZone {
    optional string timezone_id = 1;
}